### ISSUE TITLE
On .NET 6 APIs with top level statements the Swashbuckle CLI fails if host can't be initialized within 5 seconds.

### ISSUE THREAD

**Issue by jvos-priva** [domaindrivendev/Swashbuckle.AspNetCore#2355]
Using v6.2.3 of the Swashbuckle CLI to generate OpenAPI documents for a .NET 6 API using top level statements.

We have been using this for a while without issues, but after I added one line of code to our initialization, the CLI started failing consistently. This line of code was the following:

``` csharp
var builder = WebApplication.CreateBuilder(args);

// Added the following
builder.Configuration.AddAzureKeyVault(
    new Uri($"https://{builder.Configuration["keyvault:name"]}.vault.azure.net/"),
    new DefaultAzureCredential());

// Continue with other init stuff here
...
```

The error show is as follows:

```
Unhandled exception. System.InvalidOperationException: A type named 'StartupProduction' or 'Startup' could not be found in assembly 'MinimalWithSwashbuckle'.
   at Microsoft.AspNetCore.Hosting.StartupLoader.FindStartupType(String startupAssemblyName, String environmentName) in Microsoft.AspNetCore.Hosting.dll:token 0x6000136+0x14c
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.BuildCommonServices(AggregateException& hostingStartupErrors) in Microsoft.AspNetCore.Hosting.dll:token 0x6000189+0x2de
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.<>c__DisplayClass14_2.<BuildCommonServices>b__5(IServiceProvider _) in Microsoft.AspNetCore.Hosting.dll:token 0x6000289+0x0
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context) in Microsoft.Extensions.DependencyInjection.dll:token 0x6000094+0x0
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument) in Microsoft.Extensions.DependencyInjection.dll:token 0x60000a6+0x23
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context) in Microsoft.Extensions.DependencyInjection.dll:token 0x600008e+0x61
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument) in Microsoft.Extensions.DependencyInjection.dll:token 0x60000a5+0x4e
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope) in Microsoft.Extensions.DependencyInjection.dll:token 0x600008b+0x26
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType) in Microsoft.Extensions.DependencyInjection.dll:token 0x6000068+0x38
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory) in System.Collections.Concurrent.dll:token 0x60000d3+0x4f
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope) in Microsoft.Extensions.DependencyInjection.dll:token 0x6000066+0xd
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType) in Microsoft.Extensions.DependencyInjection.dll:token 0x6000061+0x0
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider) in Microsoft.Extensions.DependencyInjection.Abstractions.dll:token 0x6000094+0xe
   at Microsoft.AspNetCore.Hosting.WebHost.EnsureStartup() in Microsoft.AspNetCore.Hosting.dll:token 0x600014b+0x9
   at Microsoft.AspNetCore.Hosting.WebHost.EnsureApplicationServices() in Microsoft.AspNetCore.Hosting.dll:token 0x600014a+0x8
   at Microsoft.AspNetCore.Hosting.WebHost.Initialize() in Microsoft.AspNetCore.Hosting.dll:token 0x6000147+0x0
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build() in Microsoft.AspNetCore.Hosting.dll:token 0x6000188+0x9a
   at Swashbuckle.AspNetCore.Cli.Program.GetServiceProvider(Assembly startupAssembly) in C:\projects\ahoy\src\Swashbuckle.AspNetCore.Cli\Program.cs:line 133
   at Swashbuckle.AspNetCore.Cli.Program.<>c.<Main>b__0_3(IDictionary`2 namedArgs) in C:\projects\ahoy\src\Swashbuckle.AspNetCore.Cli\Program.cs:line 72
   at Swashbuckle.AspNetCore.Cli.CommandRunner.Run(IEnumerable`1 args) in C:\projects\ahoy\src\Swashbuckle.AspNetCore.Cli\CommandRunner.cs:line 68
   at Swashbuckle.AspNetCore.Cli.CommandRunner.Run(IEnumerable`1 args) in C:\projects\ahoy\src\Swashbuckle.AspNetCore.Cli\CommandRunner.cs:line 59
   at Swashbuckle.AspNetCore.Cli.Program.Main(String[] args) in C:\projects\ahoy\src\Swashbuckle.AspNetCore.Cli\Program.cs:line 111
```

After some sleuthing I found out the error is a red herring and instead it is caused by the following swallowed exception:
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/c7d580aa3860649de7d5faaa7b4d77dbceeaf5b7/src/Swashbuckle.AspNetCore.Cli/HostingApplication.cs#L85-L88

Which is thrown here:
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/c7d580aa3860649de7d5faaa7b4d77dbceeaf5b7/src/Swashbuckle.AspNetCore.Cli/HostFactoryResolver.cs#L259-L262

Which is caused by the use of a default timeout setting here:
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/c7d580aa3860649de7d5faaa7b4d77dbceeaf5b7/src/Swashbuckle.AspNetCore.Cli/HostFactoryResolver.cs#L22

Ideally I would have liked to see a way to specify the timeout via the CLI's command line arguments, but there does not seem to be a way to do so. At the least it would be a good idea not to swallow the exception but set it up in such a way that the application fails with an error that communicates better the source of the problem.

This might ofcourse not be easily reproducible by just adding the same line of code (it might run slightly quicker on your machine), but adding a `Thread.Sleep(6000)` should probably do the trick.

**Comment by zyofeng** [comment]
similar issue existed in NSWAG which I have put a PR that sets the timeout to 60seconds, that is also the timeout used by EF Core tools.

**Comment by SzepLaci** [comment]
When do you plan to release this change?

**Comment by psbhavsar** [comment]
Any ETA for this PR to be released? We're running "swagger & NSwag" on postbuild events to generate the API specs, and NuGet client.

**Comment by zyofeng** [comment]
This is probably a question for repo owner @domaindrivendev.

**Comment by Ruud2000** [comment]
I ran into this issue once I added configuration to setup Azure App Configuration. Also using the minimal hosting model in 6.0.

My coworker came up with the following workaround:

- Use the `EnvironmentVariables` to set the `DOTNET_ENVIRONMENT` to some name unique to for generating the swagger (e.g. `SwaggerBuild`)

```xml
  <Target Name="openapi" AfterTargets="Build">
    <Message Text="generating openapi" Importance="high" />
    <Exec EnvironmentVariables="DOTNET_ENVIRONMENT=SwaggerBuild" Command="dotnet tool run swagger tofile --yaml --output openapi.yaml $(OutputPath)$(AssemblyName).dll service" />
  </Target>
```

- In `Program.cs` you can then bypass sections that are not relevant to generate the swagger (like setting up Azure App Configuration) based on the environment name.

```c
if (builder.Environment.EnvironmentName != "SwaggerBuild")
{
    builder.Configuration.AddAzureAppConfiguration(options => options
        .Connect(new Uri(appConfigOptions.EndPoint), new DefaultAzureCredential())
        .UseFeatureFlags());
}
```

**Comment by markwalsh-liverpool** [comment]
@Ruud2000 How do you then stop the API?

**Comment by martincostello** [comment]
Fixed by #2381

**Comment by react-o-licious** [comment]
> Ruud2000's comment on the EnvironmentVariables is a good one. And can be used differently. For instance, in my app, after reading his comment, I set swagger to run on Development because I have some settings only in appsettings.development.json `<Exec EnvironmentVariables="DOTNET_ENVIRONMENT=Development" Command="dotnet swagger`
> 
> Generally speaking, `dotnet tool swagger` is fully starting your app, so you need to ensure the appsettings and other parameters on start are correct to be able to run the open api generation.
> 
> The exception about Startup remains very annoying since it does NOT say what is wrong. As we see here, in most cases, the error message is misleading, it has NOTHING to do with the Startup class, but just that the app could not start properly.

Cheers @ericbl and @Ruud2000 - Just like my situation, my API was throwing a runtime error but the "Startup" Exception was masking it. I should have tested before running the swagger dotnet-tool !
